### PR TITLE
Introducing aarch64 support for unikernels

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This is MR for the original review. I'm planning to make this branch as a feature branch to upstream. So it will never be merged.
This MR includes the implementation of the aarch64 architecture support for unikraft.
It allows to build elf and zboot images and run in in xen environment of aarch64.
Also supports gic implementation and generic timer support.

### CODE REVIEW

File: plat/xen/include/xen-arm/setup.h [hunk 1]
```diff
@@ -32,10 +32,18 @@
 
 #ifndef __SETUP_H__
 #define __SETUP_H__
-
```
Comment (al1img) [inline plat/xen/include/xen-arm/setup.h:35]: keep empty line

File: plat/xen/include/xen-arm/os.h [hunk 2]
```diff
@@ -108,6 +133,5 @@ static inline int irqs_disabled(void)
 }
 
 void block_domain(__snsec until);
-
```
Comment (al1img) [inline plat/xen/include/xen-arm/os.h:111]: Not relevant change.

File: plat/xen/hv_console.c [hunk 3]
```diff
@@ -106,14 +111,22 @@ static struct xencons_interface *console_ring;
 static uint32_t console_evtchn;
 static int console_ready;
 
-#ifdef CONFIG_PARAVIRT
+#if defined(CONFIG_PARAVIRT)
```
Comment (al1img) [inline plat/xen/hv_console.c:114]: Do you need this change?

File: plat/Config.uk [hunk 4]
```diff
@@ -100,7 +100,6 @@ config FPSIMD
 
 menuconfig LIBGIC
 	bool "Arm GIC (generic interrupt controller) support"
-	select LIBOFW
```
Comment (al1img) [inline plat/Config.uk:103]: is it ok to remove?

File: plat/xen/Makefile.uk [hunk 5]
```diff
@@ -13,11 +13,13 @@ $(eval $(call addplatlib_s,xen,libxenbus,$(CONFIG_XEN_XENBUS)))
...
 $(eval $(call addplatlib_s,xen,libxenblkfront,$(CONFIG_XEN_BLKFRONT)))
 $(eval $(call addplatlib_s,xen,libxen9pfront,$(CONFIG_XEN_9PFRONT)))
+$(eval $(call addplatlib_s,xen,libxengic,$(CONFIG_LIBGIC)))
+$(eval $(call addplatlib_s,xen,libxenofw,$(CONFIG_LIBOFW)))
 
 ##
 ## Xen platform compilation settings
 ##
-LIBXENPLAT_ASFLAGS-y     += -DXENPLAT -D__XEN_INTERFACE_VERSION__=$(XEN_INTERFACE_VERSION)
+LIBXENPLAT_ASFLAGS-y     += -DXENPLAT -D__XEN_INTERFACE_VERSION__=$(XEN_IFACE_VERSION)
```
Comment (al1img) [inline plat/xen/Makefile.uk]: Is it IFACE for asm?

File: plat/common/arm/time.c [hunk 6]
```diff
@@ -50,6 +50,13 @@ static const char * const arch_timer_list[] = {
 	NULL
 };
 
+#if defined(CONFIG_PLAT_XEN)
```
Comment (al1img) [inline plat/common/arm/time.c:53]: Commit message: remove comma after arch_timer.

File: plat/xen/Config.uk [hunk 7]
```diff
@@ -1,14 +1,14 @@
 menuconfig PLAT_XEN
        bool "Xen guest image"
        default n
-       depends on ((ARCH_X86_32 && !HAVE_SMP) || (ARCH_X86_64 && !HAVE_SMP) || (ARCH_ARM_32 && !HAVE_SMP))
```
[the remark this answers is not in the document]
Comment (oleksiimoisieiev) [reply plat/xen/Config.uk:4]: fixed

File: plat/xen/include/xen-arm/setup.h (hunk 1 above)
Comment (oleksiimoisieiev) [reply plat/xen/include/xen-arm/setup.h:35]: fixed

File: plat/xen/include/xen-arm/os.h (hunk 2 above)
Comment (oleksiimoisieiev) [reply plat/xen/include/xen-arm/os.h:111]: fixed

File: plat/xen/hv_console.c (hunk 3 above)
Comment (oleksiimoisieiev) [reply plat/xen/hv_console.c:114]: fixed

File: plat/xen/Makefile.uk (hunk 5 above)
Comment (oleksiimoisieiev) [reply plat/xen/Makefile.uk]: Updated.

File: plat/xen/include/xen/xen-compat.h [hunk 8]
```diff
@@ -29,11 +29,16 @@
 
 #define __XEN_LATEST_INTERFACE_VERSION__ 0x00040900
 
+#define DO_EXPAND(VAL) (VAL##1)
```
Comment (al1img) [inline plat/xen/include/xen/xen-compat.h:32]: Commit message `being empty`.

Comment (oleksiimoisieiev) [reply plat/xen/include/xen/xen-compat.h:32]: fixed

File: plat/common/arm/time.c (hunk 6 above)
Comment (oleksiimoisieiev) [reply plat/common/arm/time.c:53]: fixed

**Review by al1img — APPROVED** [review]

**Pull request merged**